module: expose `getPackageJSON` utility (#55229)
nodejs/node: doc/api/module.md

* Returns: {string|undefined} Path to the [module compile cache][] directory if it is enabled,
  or `undefined` otherwise.

### `module.getPackageJSON(startPath[, everything])`

<!-- YAML

> added: VERSION

Author: What keyword am I suppose to put here? I can't find it in the contributing docs (ex `api-documentation.md`)

Reviewer: it should be replaceme all capital.

-->

> Stability: 1.1 - Active Development

> * `startPath` {URL\['pathname']} Where to start looking

Reviewer 2: ```suggestion
* `startPath` {string} Where to start looking
```

> * `everything` {boolean} Whether to return the full contents of the found package.json

Reviewer 3: I'm not sure it is a good idea to introduce yet-another way of loading a JSON file, wouldn't it be better to return the path and let the user deal with reading/parsing the file?

Author: I considered that (there's another PR I opened and closed that exposed `findPackageJSON`, which did exactly that). We're already reading the pjson and caching it, so IMO that would force a de-op as well as force the user to manually do what we're already doing. So, why?

Reviewer 4: I think returning the parsed data means every time we add consider parsing another field in the package.json, it needs to be surfaced into this API, even though that field may not even be useful for tools, but we'll end up paying the serialization/deserialization cost even though internally only selected places need selected fields. IMO even just returning the string would be better than this. Users need this for the lookup algorithm, they tend to have other fields to parse anyway.

Author: Sorry, @joyeecheung I'm not sure I follow in your message. I think you've misunderstood the behaviour here: internally, we do not use the full version—we do use `getNearestParentPackageJSON` but do not set `everything` to `true`. When `everything` is not `true`, only the select fields we use internally are parsed. That has not changed in this PR.

Author: > May I recommend a different naming and API surface?
> 
> `module.findPackageJson(path, { includeContents: true })`

@anonrig if you recall, when we were working on this a few days ago, that did not make sense due to how the C++ worked, which is what caused me to abandon the `findPackageJSON` + `getPackageJSON` route in the first place.

Regarding "find" + `includeContents`, this seems counter-intuitive to me: I expect something called "find" to only locate, not retrieve. I could perhaps see the inverse

`module.getPackageJSON(path, { includeContents: false })`

I do not like either option though because the shape of the return is changed significantly in a foot-gun way:

```
getPackageJSON(path, false) // '…/package.json'

getPackageJSON(path, true)  // { data: {…}, path: '…/package.json' }
```
(I simplified the 2nd arg for brevity, not necessarily to say we shouldn't use an object)

I suppose it could always return an object for consistency, but that seems like a clumsy compromise

```
getPackageJSON(path, false) // { path: '…/package.json' }

getPackageJSON(path, true)  // { data: {…}, path: '…/package.json' }
```

If we were to do this, I think exposing 2 different utils (`findPackageJSON` + `getPackageJSON`) would be better.

> * `everything` {boolean} Whether to return the full contents of the found package.json
> * Returns: {undefined | {
>   data: {
>   name?: string,
>   type?: 'commonjs' | 'module' | 'none',
>   exports?: string | string\[] | Record\<string, unknown>,
>   imports?: string | string\[] | Record\<string, unknown>,
>   \[key: string]?: unknown,
>   },
>   path: URL\['pathname'],
>   }}

Reviewer 2: ```suggestion
* `everything` {boolean} Whether to return the full contents of the found package.json
* Returns: {Object|undefined} The `package.json` contents.
  * data {Object} An object containing the module metadata with the following schema:
    * name {string|undefined} The name of the module.
    * type {string|undefined} The type of module, either 'commonjs', 'module', or 'none'.
    ...
```

Something like this is now objects are typically described throughout the docs.

> In addition to being available to users, this utility is used internally when

Reviewer 4: I don't think it's a good idea to talk about what is used internally in the docs - users would assume that this can be patchable, like what you do with many other module APIs. It also hints that internally, the consumer of this function is JavaScript, which isn't necessarily the case if we port the loader to C++ again for performance.

Author: Good call :)

* Returns: {string|undefined} Path to the [module compile cache][] directory if it is enabled,
  or `undefined` otherwise.

### `module.getPackageJSON(startLocation[, everything])`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.1 - Active Development

* `startLocation` {string} A fully resolved FS path or file URL to start looking
* `everything` {boolean} Whether to return the full contents of the found package.json

> * Returns: {Object | undefined}

Reviewer 2: ```suggestion
* Returns: {Object|undefined}
```
Ditto to the rest

Author: why?

Reviewer 2: It's the way the docs are formatted. There is no space between the `|`.

Author: Does it actually matter or is it a style/consistency issue?

Reviewer 2: I don't think the linter will throw an error, but it's documented in the Style Guide (https://github.com/nodejs/node/blob/main/doc/README.md#function-arguments-and-returns)

Reviewer 3: Yeah let's keep them consistent, i.e. without spaces

  * data: {Object}
    * name: {string | undefined}
    * type: {string | undefined}

>     * exports: string | string\[] | Record\<string, unknown> | undefined

Reviewer 2: Is this a valid type?

Author: I dunno if the validator considers it one, but it absolutely valid 🙂

Reviewer 3: Note that we are not using TS nor JSDoc syntax, so it doesn't really make sense to claim it's "valid syntax" if the validator disagrees
```suggestion
    * exports: {string|string[]|Object|undefined}
```

    * imports: string | string\[] | Record\<string, unknown> | undefined
    * …
  * path: {string}

> Retreives the contents and location of the package.json closest to the supplied `startLocation`;

Reviewer 3: ```suggestion
Retreives the contents and location of the `package.json` closest to the supplied `startLocation`;
```

> this behaves identically to node's own lookup and consumption of package.json for a given module.

Reviewer 3: ```suggestion
this behaves identically to Node.js own lookup and consumption of `package.json` for a given module.
```
